File: packages/mobile/modules/stella-cloud-browser-session/android/src/main/java/com/stella/cloudbrowsersession/StellaCloudBrowserSessionModule.kt
package com.stella.cloudbrowsersession

import android.webkit.CookieManager
import androidx.webkit.CookieManagerCompat
import androidx.webkit.WebViewFeature
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.net.URI
import java.net.URISyntaxException
import java.nio.charset.StandardCharsets
import java.time.Instant
import java.util.Locale

private const val MAXIMUM_COOKIES = 128
private const val MAXIMUM_BYTES = 32 * 1024

class StellaCloudBrowserSessionModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("StellaCloudBrowserSession")

    Function("isCaptureAvailable") {
      WebViewFeature.isFeatureSupported(WebViewFeature.GET_COOKIE_INFO)
    }

    AsyncFunction("captureCookies") Coroutine { rawUrl: String ->
      if (!WebViewFeature.isFeatureSupported(WebViewFeature.GET_COOKIE_INFO)) {
        throw CaptureUnavailableException()
      }
      val uri = validatedHttpsUri(rawUrl)
      val rawCookies = CookieManagerCompat.getCookieInfo(CookieManager.getInstance(), uri.toString())
      val nowEpochSeconds = Instant.now().epochSecond
      var totalBytes = 0
      val result = mutableListOf<Map<String, Any>>()

      rawCookies.forEach { rawCookie ->
        val cookie = CloudBrowserCookieParser.parse(rawCookie, uri, nowEpochSeconds)
          ?: return@forEach
        if (result.size >= MAXIMUM_COOKIES) throw SessionTooLargeException()
        totalBytes += listOf(
          cookie.name,
          cookie.value,
          cookie.domain,
          cookie.path,
          cookie.sameSite,
        ).sumOf { field -> field.toByteArray(StandardCharsets.UTF_8).size }
        if (totalBytes > MAXIMUM_BYTES) throw SessionTooLargeException()
        result.add(cookie.toMap())
      }
      result
    }
  }

  private fun validatedHttpsUri(rawUrl: String): URI {
    val uri = try {
      URI(rawUrl)
    } catch (_: URISyntaxException) {
      throw InvalidOriginException()
    }
    if (
      uri.scheme?.lowercase(Locale.US) != "https" ||
      uri.host.isNullOrBlank() ||
      uri.userInfo != null
    ) {
      throw InvalidOriginException()
    }
    return uri
  }
}

private class InvalidOriginException : CodedException(
  "invalid_origin",
  "A valid HTTPS URL is required.",
  null,
)

private class CaptureUnavailableException : CodedException(
  "capture_unavailable",
  "This Android WebView cannot export complete cookie attributes.",
  null,
)

private class SessionTooLargeException : CodedException(
  "session_too_large",
  "The site session is too large to transfer.",
  null,
)

File: packages/mobile/modules/stella-cloud-browser-session/ios/StellaCloudBrowserSessionModule.swift
import ExpoModulesCore
import Foundation
import WebKit

public final class StellaCloudBrowserSessionModule: Module {
  private static let maximumCookies = 128
  private static let maximumBytes = 32 * 1024

  public func definition() -> ModuleDefinition {
    Name("StellaCloudBrowserSession")

    AsyncFunction("captureCookies") { (rawURL: String) async throws -> [[String: Any]] in
      guard
        let url = URL(string: rawURL),
        url.scheme == "https",
        url.user == nil,
        url.password == nil,
        let hostname = url.host?.lowercased()
      else {
        throw Exception(name: "invalid_origin", description: "A valid HTTPS URL is required.")
      }

      let allCookies = await WKWebsiteDataStore.default().httpCookieStore.allCookies()
      var totalBytes = 0
      var result: [[String: Any]] = []

      for cookie in allCookies where Self.applies(cookie: cookie, to: url, hostname: hostname) {
        guard result.count < Self.maximumCookies else {
          throw Exception(name: "session_too_large", description: "The site session contains too many cookies.")
        }
        let sameSite = Self.sameSite(cookie)
        let fields = [cookie.name, cookie.value, cookie.domain, cookie.path, sameSite]
        totalBytes += fields.reduce(0) { count, value in
          count + value.lengthOfBytes(using: .utf8)
        }
        guard totalBytes <= Self.maximumBytes else {
          throw Exception(name: "session_too_large", description: "The site session is too large to transfer.")
        }
        result.append([
          "name": cookie.name,
          "value": cookie.value,
          "domain": cookie.domain,
          "path": cookie.path.isEmpty ? "/" : cookie.path,
          "expires": cookie.expiresDate?.timeIntervalSince1970 ?? -1,
          "httpOnly": cookie.isHTTPOnly,
          "secure": cookie.isSecure,
          "sameSite": sameSite,
        ])
      }
      return result
    }
  }

  private static func applies(cookie: HTTPCookie, to url: URL, hostname: String) -> Bool {
    if let expires = cookie.expiresDate, expires.timeIntervalSinceNow <= 0 {
      return false
    }
    let domain = cookie.domain.lowercased().trimmingCharacters(in: CharacterSet(charactersIn: "."))
    guard hostname == domain || hostname.hasSuffix("." + domain) else {
      return false
    }
    return !cookie.isSecure || url.scheme == "https"
  }

  private static func sameSite(_ cookie: HTTPCookie) -> String {
    let key = HTTPCookiePropertyKey("SameSite")
    guard let raw = cookie.properties?[key] as? String else {
      return "Lax"
    }
    switch raw.lowercased() {
    case "strict": return "Strict"
    case "none": return "None"
    default: return "Lax"
    }
  }
}

private extension WKHTTPCookieStore {
  func allCookies() async -> [HTTPCookie] {
    await withCheckedContinuation { continuation in
      getAllCookies { cookies in
        continuation.resume(returning: cookies)
      }
    }
  }
}
